### PR TITLE
update to Geant4 10, switch to Shielding list plus optics

### PR DESCRIPTION
**Pull request by mastbaum**
Updated for compatibility with Geant4 10.0 and the latest, crankiest
Apple compiler:

* Fixed missing includes and namespace issues for CLHEP PhysicalConstants
  and SystemOfUnits constants
* Fixed shadowing of members and locals throughout

The existing physics list needed some work anyway, and we've been
discussing switching to the built-in Shielding physics list, so RAT's
physics list has been replaced with a Shielding subclass that adds
optical photon physics, namely:

* Geant4 G4Cerenkov
* Geant4 G4OpBoundaryProcess
* RAT GLG4Scint
* RAT GLG4OpAttenuation

I did some minimal testing with the new physics list and detect a reasonable number of scintillation photons, etc.; should do more thorough validation before merging.

### CODE REVIEW

File: src/daq/PDFPMTTime.cc [hunk 1]
```diff
@@ -45,7 +45,7 @@ PDFPMTTime::PDFPMTTime(string pmt_model) {
 
 PDFPMTTime::~PDFPMTTime() {}
 
-double PDFPMTTime::PickTime(double time) const {
+double PDFPMTTime::PickTime(double /*time*/) const {
```
Comment (BenLand100) [inline src/daq/PDFPMTTime.cc]: This bug is fixed in #23

**Comment by BenLand100** [comment]
Besides collisions with other PRs this all looks good from a code standpoint. It looks like all cases of shadowed variables were properly refactored. Presumably the idea is to merge this, then update #29 to reflect additions to this GEANT4.10ified RAT?

**Comment by mastbaum** [comment]
Thanks for taking a look. That was indeed what I had in mind -- I have a nearly-working version of the new code from #20 merged into this... just trying to sort out some apparent shadowing issues with ROOT + Geant4 in order to preserve `-Werror`.

**Comment by orebi** [comment]
To be fair, since we don't have a full validation of rat-pac with the previous GEANT I would propose that we merge this and validate after the fact. Do we have any reason to suppose that the previous version is any closer to accurate?

**Comment by mastbaum** [comment]
I've no issue with merging then validating, though it would be useful for anyone who's actively using RAT-PAC to try this branch and see if anything's gone terribly wrong. The previous version is based on MiniCLEAN RAT (with some omissions).

**Comment by MarcFBergevin** [comment]
(This was also sent as an email)

Hi Andy,

Thank you. I decided to take another approach, since I was most definitively doing things the wrong way. I was getting the code from your fork instead of as the pull request. I am including the step I am currently doing in case since there is still an issue here:

> git remote -v
fork    git@github.com:MarcFBergevin/rat-pac.git (fetch)
fork    git@github.com:MarcFBergevin/rat-pac.git (push)
main    https://github.com/rat-pac/rat-pac.git (push)
main    https://github.com/rat-pac/rat-pac.git (fetch)
...

>  git fetch main pull/29/head:GV10

> git checkout GV10 

>./configure
>source env.sh
>scons -c
>rm -rf include/RAT
>scons
...
/Users/marcbergevin/root/root-v5-34-30/include src/gen/DecayChain_Gen.cc
src/gen/DecayChain_Gen.cc:82:35: error: no member named 'GetIon' in 'G4ParticleTable'
            particleDef = particleTable->GetIon(Z, A, excitationEnergy);
                          ~~~~~~~~~~~~~  ^
1 error generated.
scons: *** [build/macosx64/gen/DecayChain_Gen.o] Error 1

>lt src/gen/DecayChain_Gen*
...  2.3K Oct 30  2014 src/gen/DecayChain_Gen.hh
...  7.0K Oct 30  2014 src/gen/DecayChain_Gen.cc

So doing this, there is this function that would need to be fixed. Looking at the Geant4.10 release notes:

Hadronic physics

    The CHIPS package, the parameterized (Gheisha-like) models LEP/HEP and the isotope_production module have been removed.
    GetIon() and FindIon() methods are removed from G4ParticleTable class. These methods have moved to G4IonTable class with some enhancements.

Doing the following grep command:

> grep GetIon -r *
src/gen/DecayChain_Gen.cc:        particleDef = particleTable->GetIon(Z, A, excitationEnergy);
src/gen/GLG4VertexGen.cc:      G4ParticleTable::GetParticleTable()->GetIon(Z, A, 0.0);
src/gen/GLG4VertexGen.cc:      G4ParticleTable::GetParticleTable()->GetIon(Z, A, 0.0);
src/gen/GLG4VertexGen.cc:    G4ParticleTable::GetParticleTable()->GetIon(Z, A, 0.0);
src/gen/VertexGen_Spectrum.cc:                        newTestGunG4Code=G4ParticleTable::GetParticleTable()->GetIon(Z, A, 0.0);
src/gen/VertexGen_WIMP.cc:      fNucleus = G4ParticleTable::GetParticleTable()->GetIon(18, 40, 0.0);
src/gen/VertexGen_WIMP.cc:      fNucleus = G4ParticleTable::GetParticleTable()->GetIon(10, 20, 0.0);


Since I do not know enough about git and I am still figuring things out, is it possible that there is code from another branch that is leaking in to this one? Should I do these tests in a new environment?

Thank you for your help,
Marc

**Comment by mastbaum** [comment]
It appears that this actually changed between G4 versions 10.00 and 10.1: in my 10.00.p02 `G4ParticleTable` still has `GetIon` methods with a comment warning of future deprecation. So it looks like these calls you identified will indeed have to be changed.

FWIW I've not used your method for checking out a pull request branch, but the following works for me with G4 10.00.p02:
```
$ source ~/sw/root-5.34.18/bin/thisroot.sh
$ source ~/sw/geant4.10.00.p02/bin/geant4.sh
$ git clone git@github.com:mastbaum/rat-pac rat-pac-g4v10
$ cd rat-pac-g4v10
$ git checkout geant4-v10-compat
$ ./configure && source env.sh
$ scons
```

**Comment by MarcFBergevin** [comment]
Thank you. I did a quick change from particleTable to IonTable. The code did compile, however it crashed in execution. I was hopeful that the situation would be that easy, but it is apparently not. 

I need to finish some simulation work today, but should be available next Friday to debug in greater details.

File: src/db/json.hh [hunk 2]
```diff
@@ -182,10 +182,10 @@ namespace json {
             static void wrongType(Type actual, Type requested);
         
             // Throws a runtime_error if the type of the Value does not match the given Type
-            inline void checkType(Type type) const { if (this->type != type) { wrongType(this->type,type); } } 
+            inline void checkType(Type _type) const { if (this->type != _type) { wrongType(this->type, _type); } }
```
Comment (mastbaum) [inline src/db/json.hh]: @BenLand100 for this and similar, I was getting e.g.
```
In file included from rat.cc:23:
In file included from include/RAT/Log.hh:87:
include/RAT/json.hh:185:40: error: declaration shadows a field of 'json::Value' [-Werror,-Wshadow]
            inline void checkType(Type type) const { if (this->type != type) { wrongType(this->type,type); } }
                                       ^
include/RAT/json.hh:203:18: note: previous declaration is here
            Type type;
                 ^
```

**Comment by mastbaum** [comment]
@MarcFBergevin @BenLand100 I've updated for G4 10.01, changing `G4ParticleTable::GetIon` calls to `GetIonTable::GetIon` per the [10.00 release notes](http://geant4.cern.ch/support/ReleaseNotes4.10.0.html#3.).

Also a couple of other small tweaks to get things compiling on my platform (Apple LLVM version 6.0 (clang-600.0.57)).

File: src/db/json.hh (hunk 2 above)
Comment (BenLand100) [inline src/db/json.hh]: shadow is my least favorite warning. I was under impression that the scoping was unambiguous and the explicit ```this->``` was also unambiguous at allowing access to the "shadowed" object variables, so it's unclear why this wasn't a warning before and is recently (compilers!?!?!), but this popped up just before fastjson was merged into snoplus rat and similar changes were made there.

**Comment by BenLand100** [comment]
I believe we decided on the call to go ahead and merge this. Were there any further concerns? Otherwise I'll press the big green button.

**Comment by MarcFBergevin** [comment]
No concerns on my part.

**Pull request merged**